src/protocols/pci_io.rs

```diff
@@ -0,0 +1,200 @@
+pub const ATTRIBUTE_DUAL_ADDRESS_CYCLE: Attribute = 0x00008000;
+pub const ATTRIBUTE_ISA_IO_16: Attribute = 0x00010000;
+pub const ATTRIBUTE_VGA_PALETTE_IO_16: Attribute = 0x00020000;
+pub const ATTRIBUTE_VGA_IO_16: Attribute = 0x00040000;
+
+pub type AttributeOperation = u32;
+
+pub const ATTRIBUTE_OPERATION_GET: AttributeOperation = 0x00000000;
+pub const ATTRIBUTE_OPERATION_SET: AttributeOperation = 0x00000001;
+pub const ATTRIBUTE_OPERATION_ENABLE: AttributeOperation = 0x00000002;
+pub const ATTRIBUTE_OPERATION_DISABLE: AttributeOperation = 0x00000003;
+pub const ATTRIBUTE_OPERATION_SUPPORTED: AttributeOperation = 0x00000004;
+
+pub const PASS_THROUGH_BAR: u8 = 0xff;
+
+pub type ProtocolPollIoMem = eficall! {fn(
+    *mut Protocol,
+    Width,
+    u8,
+    u64,
+    u64,
+    u64,
+    u64,
+    *mut u64,
+) -> crate::base::Status};
+
+pub type ProtocolIoMem = eficall! {fn(
+    *mut Protocol,
+    Width,
+    u8,
+    u64,
+    usize,
+    *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolConfig = eficall! {fn(
+    *mut Protocol,
+    Width,
+    u32,
+    usize,
+    *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolCopyMem = eficall! {fn(
+    *mut Protocol,
+    Width,
+    u8,
+    u64,
+    u8,
+    u64,
+    usize,
+) -> crate::base::Status};
+
+pub type ProtocolMap = eficall! {fn(
+    *mut Protocol,
+    Operation,
+    *mut core::ffi::c_void,
+    *mut usize,
+    *mut crate::base::PhysicalAddress,
+    *mut *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolUnmap = eficall! {fn(
+    *mut Protocol,
+    *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolAllocateBuffer = eficall! {fn(
+    *mut Protocol,
+    crate::system::AllocateType,
+    crate::system::MemoryType,
+    usize,
+    *mut *mut core::ffi::c_void,
+    Attribute,
+) -> crate::base::Status};
+
+pub type ProtocolFreeBuffer = eficall! {fn(
+    *mut Protocol,
+    usize,
+    *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolFlush = eficall! {fn(
+    *mut Protocol,
+) -> crate::base::Status};
+
+pub type ProtocolGetLocation = eficall! {fn(
+    *mut Protocol,
+    *mut usize,
+    *mut usize,
+    *mut usize,
+    *mut usize,
+) -> crate::base::Status};
+
+pub type ProtocolAttributes = eficall! {fn(
+    *mut Protocol,
+    AttributeOperation,
+    Attribute,
+    *mut Attribute,
+) -> crate::base::Status};
+
+pub type ProtocolGetBarAttributes = eficall! {fn(
+    *mut Protocol,
+    u8,
+    *mut Attribute,
+    *mut *mut core::ffi::c_void,
+) -> crate::base::Status};
+
+pub type ProtocolSetBarAttributes = eficall! {fn(
+    *mut Protocol,
+    Attribute,
+    u8,
+    *mut u64,
+    *mut u64,
+) -> crate::base::Status};
+
+#[repr(C)]
+pub struct Access {
+    pub read: ProtocolIoMem,
+    pub write: ProtocolIoMem,
+}
+
+#[repr(C)]
+pub struct ConfigAccess {
+    pub read: ProtocolConfig,
+    pub write: ProtocolConfig,
+}
```

Comment:
At some point we need to agree which traits to auto-derive for ops. I left this as is, but I think we should add `Clone, Debug` to all repr-C structs.